README
NAME
    Business::ID::SIM - Validate Indonesian driving license number (nomor
    SIM)

VERSION
    This document describes version 0.05 of Business::ID::SIM (from Perl
    distribution Business-ID-SIM), released on 2014-08-16.

SYNOPSIS
     use Business::ID::SIM qw(parse_sim);

     my $res = parse_sim(sim => "0113 40 00 0001");

DESCRIPTION
    This module can be used to validate Indonesian driving license number,
    Nomor Surat Izin Mengemudi (SIM).

    SIM is composed of 12 digits as follow:

     yymm.pp.aa.ssss

    yy and mm are year and month of birth, pp and aa are area code
    (province+district of some sort), ssss is 4-digit serial most probably
    starting from 1.

    Note that there are several kinds of SIM (A, B1, B2, C) but this is not
    encoded in the SIM number and all SIM's have the same number.

FUNCTIONS
  parse_sim(%args) -> [status, msg, result, meta]
    Validate Indonesian driving license number (nomor SIM).

    Arguments ('*' denotes required arguments):

    *   sim* => *str*

        Input to be parsed.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

     (any)

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Business-ID-SIM>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-Business-ID-SIM>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Business-ID-SIM>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Business-ID-SIM

0.05     2014-08-16 (SHARYANTO) - Happy 19th CPAN Day!

         - Add CLI script(s).


0.04    2013-12-04 (SHARYANTO)

        - No functional changes. Use Data::Clean::JSON 0.11's get_cleanser().


0.03    2013-09-09 (SHARYANTO)

        [INCOMPATIBLE CHANGES]

        - Replace old interface (validate_sim() + OO interface) with Rinci-based
          parse_sim().


0.02    2010-02-26 (SHARYANTO)

        - Generate META.yml (using Dist::Zilla)


0.01    2009-10-06 (SHARYANTO)

        - First version, released on an unsuspecting world.


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Validate Indonesian driving license number (nomor SIM)",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Business-ID-SIM",
  "EXE_FILES" => [
    "bin/parse-sim"
  ],
  "LICENSE" => "perl",
  "NAME" => "Business::ID::SIM",
  "PREREQ_PM" => {
    "DateTime" => 0,
    "Perinci::CmdLine::Any" => "0.06",
    "Rinci" => "v1.1.0"
  },
  "TEST_REQUIRES" => {
    "Data::Clean::JSON" => "0.11",
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Clean::JSON" => "0.11",
  "DateTime" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Perinci::CmdLine::Any" => "0.06",
  "Rinci" => "v1.1.0",
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Business/ID/SIM.pm
package Business::ID::SIM;

use 5.010001;
use warnings;
use strict;

use DateTime;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw(parse_sim);

our $VERSION = '0.05'; # VERSION

# legend: S = lack of samples

# less organized than the convoluted code in NIK

my %provinces = (
    '06' => "Nanggroe Aceh Darussalam",
    '07' => "Sumatera Utara",
    '08' => "Sumatera Barat",
    '09' => "Riau, Kepulauan Riau",
    '10' => undef, # S
    '11' => "Sumatera Selatan",
    '12' => "DKI Jakarta", # most frequently encountered
    '13' => "Jawa Barat",
    '14' => "Jawa Tengah, DI Yogyakarta",
    '15' => "Jawa Timur",
    '16' => "Bali",
    '17' => "Kalimantan Timur",
    '18' => "Kalimantan Selatan",
    '19' => "Sulawesi Selatan",
    '20' => "Sulawesi Utara, Gorontalo",
    '21' => undef, # S
    '22' => 'Papua',
    '23' => 'Kalimantan Tengah',
    '24' => 'Sulawesi Tengah',
    '25' => 'Lampung',
    #'30' => "Banten", # ?, SS

    );

our %SPEC;

$SPEC{parse_sim} = {
    v => 1.1,
    summary => 'Validate Indonesian driving license number (nomor SIM)',
    args => {
        sim => {
            summary => 'Input to be parsed',
            schema => 'str*',
            pos => 0,
            req => 1,
        },
    },
};
sub parse_sim {
    my %args = @_;

    my $sim = $args{sim} or return [400, "Please specify sim"];
    my $res = {};

    $sim =~ s/\D+//g;
    return [400, "Not 12 digit"] unless length($sim) == 12;

    $res->{prov_code} = substr($sim, 4, 2);
    return [400, "Unknown province code"] unless $provinces{$res->{prov_code}};
    $res->{area_code} = substr($sim, 4, 4);

    my ($y, $m) = $sim =~ /^(..)(..)/;
    my $today = DateTime->today;
    $y += int($today->year / 100) * 100;
    $y -= 100 if $y > $today->year;
    eval { $res->{dob} = DateTime->new(day=>1, month=>$m, year=>$y) };
    return [400, "Invalid year and month of birth: $y, $m"] if $@;
    $res->{serial} = $sim =~ /(....)$/;
    return [400, "Serial starts from 1, not 0"] if $res->{serial} < 1;

    [200, "OK", $res];
}

1;
# ABSTRACT: Validate Indonesian driving license number (nomor SIM)

__END__

=pod

=encoding UTF-8

=head1 NAME

Business::ID::SIM - Validate Indonesian driving license number (nomor SIM)

=head1 VERSION

This document describes version 0.05 of Business::ID::SIM (from Perl distribution Business-ID-SIM), released on 2014-08-16.

=head1 SYNOPSIS

 use Business::ID::SIM qw(parse_sim);

 my $res = parse_sim(sim => "0113 40 00 0001");

=head1 DESCRIPTION

This module can be used to validate Indonesian driving license number, Nomor
Surat Izin Mengemudi (SIM).

SIM is composed of 12 digits as follow:

 yymm.pp.aa.ssss

yy and mm are year and month of birth, pp and aa are area code
(province+district of some sort), ssss is 4-digit serial most probably starting
from 1.

Note that there are several kinds of SIM (A, B1, B2, C) but this is not encoded
in the SIM number and all SIM's have the same number.

=head1 FUNCTIONS


=head2 parse_sim(%args) -> [status, msg, result, meta]

Validate Indonesian driving license number (nomor SIM).

Arguments ('*' denotes required arguments):

=over 4

=item * B<sim>* => I<str>

Input to be parsed.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.

 (any)

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Business-ID-SIM>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Business-ID-SIM>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Business-ID-SIM>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
